### PR TITLE
[FEATURE] Learned Step Size Quantization

### PR DESCRIPTION
**Pull request by FreakieHuang**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Add LSQ QAT algorithm. 

## Modification

1. Related LSQ components
2. SOTA experiment results
3. Comprehensive UTs. 

## BC-breaking (Optional)

Does the modification introduce changes that break the backward compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here and update the documentation.

## Checklist

**Before PR**:

- [ ] Pre-commit or other linting tools are used to fix the potential lint issues.
- [ ] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] The modification is covered by complete unit tests. If not, please add more unit tests to ensure the correctness.
- [ ] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects, like MMDet or MMSeg.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: mmrazor/models/observers/lsq_observer.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+            self.min_val, self.max_val = torch._aminmax(y, 1)
+
+        return x
+
+    def calculate_qparams(self):
+        scale = 2 * self.tensor_norm / math.sqrt(self.quant_max)
+        zero_point = torch.zeros_like(self.tensor_norm)
+        sync_tensor(scale)
+        sync_tensor(zero_point)
+        if self.pot_scale:
```
Comment (humu789) [inline mmrazor/models/observers/lsq_observer.py:55]: self.is_pot_scale

File: mmrazor/models/fake_quants/lsq.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+    @staticmethod
+    def forward(ctx, x, scale, zero_point, ch_axis, quant_min, quant_max,
+                grad_factor):
+        return _fake_quantize_learnable_per_channel_affine_training(
+            x, scale, zero_point, ch_axis, quant_min, quant_max, grad_factor)
+
+    @staticmethod
+    def symbolic(g, x, scale, zero_point, ch_axis, quant_min, quant_max,
+                 grad_factor):
+        return g.op(
```
Comment (humu789) [inline mmrazor/models/fake_quants/lsq.py:131]: custom onnx op need to be registered

File: configs/quant/qat/iao_resnet50_8xb16_cifar10.py [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+    _delete_=True,
+    _scope_='mmrazor',
+    type='FXModelWrapper',
+    model={{_base_.model}},
+    customed_skipped_method=[
+        'mmcls.models.ClsHead._get_predictions',
+        'mmcls.models.ClsHead._get_loss'
+    ])
+
+custom_hooks = [dict(type='QuantitiveHook')]
```
Comment (humu789) [inline configs/quant/qat/iao_resnet50_8xb16_cifar10.py]: This file can move configs/quantization/qat

File: mmrazor/models/task_modules/fx/graph_module.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+from typing import Any, Dict, Union
+
+import torch
+import torch.fx
+from mmengine.model import BaseModel
+from torch.fx import Graph
+from torch.fx.graph_module import GraphModule, _copy_attr
+
+
+class MMGraphModule(GraphModule, BaseModel):
```
Comment (humu789) [inline mmrazor/models/task_modules/fx/graph_module.py]: task_modules/fx and task_modules/tracer need to be unified

**Comment by FreakieHuang** [comment]
change to https://github.com/open-mmlab/mmrazor/pull/346

**Pull request closed without merging**